server/src/proxy/hotspot.rs

```diff
@@ -0,0 +1,432 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! hotspot recorder
+use std::{fmt::Write, sync::Arc, time::Duration};
+
+use ceresdbproto::storage::{
+    PrometheusQueryRequest, RequestContext, SqlQueryRequest, WriteRequest,
+};
+use common_util::{runtime::Runtime, timed_task::TimedTask};
+use log::{info, warn};
+use serde::{Deserialize, Serialize};
+use spin::Mutex;
+pub use spin::Mutex as SpinMutex;
+use tokio::sync::mpsc::{self, Sender};
+
+use crate::proxy::{hotspot_lru::HotspotLru, util};
+
+type QueryKey = String;
+type WriteKey = String;
+const TAG: &str = "hotspot autodump";
+const RECODER_CHANNEL_CAP: usize = 64 * 1024;
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(default)]
+pub struct Config {
+    /// Max items size for query hotspot
+    query_cap: Option<usize>,
+    /// Max items size for write hotspot
+    write_cap: Option<usize>,
+    dump_interval: Duration,
+    auto_dump: bool,
+    /// Max items for dump hotspot
+    auto_dump_len: usize,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            query_cap: Some(10_000),
+            write_cap: Some(10_000),
+            dump_interval: Duration::from_secs(5),
+            auto_dump: true,
+            auto_dump_len: 10,
+        }
+    }
+}
+
+enum Message {
+    Query(QueryKey),
+    // (WriteKey, row_count, field_count)
+    Write(WriteKey, usize, usize),
+}
+
+#[derive(Clone)]
+pub struct HotspotRecorder {
+    tx: Arc<Sender<Message>>,
+    stat: HotspotStat,
+}
+
+#[derive(Clone)]
+pub struct HotspotStat {
+    hotspot_query: Option<Arc<SpinMutex<HotspotLru<QueryKey>>>>,
+    hotspot_write: Option<Arc<SpinMutex<HotspotLru<WriteKey>>>>,
+    hotspot_field_write: Option<Arc<SpinMutex<HotspotLru<WriteKey>>>>,
+}
+
+impl HotspotStat {
+    /// return read count / write row count / write field count
+    pub fn dump(&self) -> Dump {
+        let format_hots = |hots: Vec<(String, u64)>| {
+            hots.into_iter()
+                .map(|(k, v)| format!("metric={k}, heats={v}"))
+                .collect()
+        };
```

Comment:
Maybe we should define it as a method of the `HotspotStat`.

Reply:
Now `dump` is a method of `HotspotStat`, does you mean `HotspotRecorder`?

Reply:
What I mean is to define the closure `format_hots` rather than the `dump` method as a static method of `HotspotStat`.

Reply:
done